osdk/src/bundle/mod.rs

```diff
@@ -260,19 +272,40 @@ impl Bundle {
         }
 
         info!("Running QEMU: {:#?}", qemu_cmd);
+        let mut child = qemu_cmd.spawn().unwrap();
+        std::thread::sleep(Duration::from_secs(1)); // Wait for QEMU to start
+        let mut stream = UnixStream::connect(qemu_socket).unwrap();
 
-        let exit_status = qemu_cmd.status().unwrap();
+        // Check VM status every 0.1 seconds and break the loop if the VM is stopped.
+        while stream.write_all(b"info status\n").is_ok() {
```

Comment:
The code coverage feature is an optional one. We do not want it to affect the efficiency of the normal mode when code coverage is enabled. Similarly, we do not want to increase the complexity of the `run` method linearly with the number of possible features. So I suggest refactoring the function in the following way

1. Detect if code coverage is enabled.
2. **Avoid the runtime cost**. To do so, if the code coverage is not enabled, this `run` method should behave in the same way as before, without extra costs (e.g., creating a monitor socket or waiting for QEMU to stop in a loop). If enabled, switch to the more heavy-weight logic.
3. **Control the code complexity**. Extract all code coverage code out of the `run` method and encapsulate it into some kinds of abstraction (a function or a struct) so that readers of this `run` method are not bothered with the details of code coverage support.